### ISSUE TITLE
JSP compilation settings not honored?

### ISSUE THREAD

**Issue by mthmulders** [OpenLiberty/open-liberty#12593]
**Describe the bug**
This issue is a follow-up on https://github.com/OpenLiberty/ci.maven/issues/838.

I'm trying to have OpenLiberty compile my JSP files on startup through the Maven goal, but that didn't work. @cherylking suggested to try it without that Maven goal by setting `<jspEngine prepareJsps="0" />` and `<webContainer deferServletLoad="false" />` in **server.xml**, but that didn't work either.

**Steps to Reproduce**

Take a [sample application](github.com/mthmulders/openliberty-mvc/) and add the following to **server.xml**:

```xml
<jspEngine prepareJsps="0" scratchdir="./scratch-dir" jdkSourceLevel="18" />
<webContainer deferServletLoad="false" />
```

Start the application using `mvn liberty:dev`.

**Expected behavior**
I would expect the designated _scratchdir_ to contain class files for the JSP files in my application. This does not happen when I start the application. When I _access_ it (with a web browser), those files appear. Since `deferServletLoad` is false I would expect it to happen when the application starts, not when it is accessed.

**Diagnostic information:**
 - OpenLiberty Version:<br /> **20.0.0.6**
 - Java Version: <br />**openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.7+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode)**
 - server.xml configuration: <br />_see above_

**Comment by cherylking** [comment]
I think the sample application link should be this: https://github.com/mthmulders/openliberty-mvc/

**Comment by epj** [comment]
I suspect the precompile isn't working because your jsp files are under WEB-INF, which isn't a public location for webapp files.  Can you try moving them out from under WEB-INF and see if precompile then finds them?

**Comment by mthmulders** [comment]
You're right, @epj. When I move the JSP files to **src/main/webapp**, they're immediately compiled when the server starts. I wasn't aware of the requirement that they may not reside under **WEB-INF**.

OTOH, I feel it's not uncommon to store JSP files there, especially with frameworks like Struts, Spring Web or Krazo. I think the idea is that the user cannot invoke those views directly, if they'd know the path to the file, but only through those MVC-style frameworks.

**Comment by hearit-danny** [comment]
Any update on this? I understand that the workaround is to move JSP files to src/main/webapp, but the JEE spec says that JSPs should be in the WEB-INF folder if you don't want them publically available. 

There are situations where you want to install a WAR on a liberty server but you don't control how that WAR is is built where there are JSPs in WEB-INF folders per the convention above, and the JSP feature seems to miss them. Any change the JSP feature can support compiling JSPs from a WEB-INF location?

**Comment by epj** [comment]
As long as you access a JSP file through an indirect means, such as  <jsp:forward path=.... />  it can be under WEB-INF.  Same for includes or fragments (jspf files).   However, any JSP that will be called directly can not be under WEB-INF.   The precompile option just assumes that any jsp in an app outside of WEB-INF are callable and precompiled.   No provision to look under WEB-INF for precompile candidates, sorry.
